source/AI.cpp

```diff
@@ -702,29 +702,59 @@ void AI::Step(const PlayerInfo &player)
 			// A fighter must belong to the same government as its parent to dock with it.
 			bool hasParent = parent && parent->GetGovernment() == gov;
 			bool hasSpace = hasParent && parent->BaysFree(isFighter);
-			if(!hasParent || (!hasSpace && !Random::Int(600)) || parent->IsDestroyed()
+			if(!hasParent || (!hasSpace && !Random::Int(1200)) || parent->IsDestroyed()
 					|| parent->GetSystem() != it->GetSystem())
 			{
 				// Find a parent for orphaned fighters and drones.
 				parent.reset();
 				it->SetParent(parent);
-				vector<shared_ptr<Ship>> parentChoices;
+				
+				auto parentChoices = vector<shared_ptr<Ship>>{};
 				parentChoices.reserve(ships.size() * .1);
-				for(const auto &other : ships)
-					if(other->GetGovernment() == gov && other->GetSystem() == it->GetSystem() && !other->CanBeCarried())
+				auto reparentWith = [&](const list<shared_ptr<Ship>> otherShips) -> bool
+				{
+					for(const auto &other : otherShips)
+						if(other->GetGovernment() == gov && other->GetSystem() == it->GetSystem() && !other->CanBeCarried())
+						{
+							if(!other->IsDisabled() && other->CanCarry(*it.get()))
+							{
+								parent = other;
+								it->SetParent(other);
+								return true;
+							}
+							else
+								parentChoices.emplace_back(other);
+						}
+					return false;
+				};
+				// Mission ships should only pick ships from the same mission.
+				auto missionIt = it->IsSpecial() && !it->IsYours()
+					? find_if(player.Missions().begin(), player.Missions().end(), [&it](const Mission &m)
 					{
-						if(!other->IsDisabled() && other->CanCarry(*it.get()))
+						auto nend = m.NPCs().end();
+						return nend != find_if(m.NPCs().begin(), nend, [&it](const NPC &n)
 						{
-							parent = other;
-							it->SetParent(other);
+							auto send = n.Ships().end();
+							return send != find_if(n.Ships().begin(), send, [&it](const shared_ptr<Ship> &s)
```

Comment:
This part looks very magical to me. Could use some comments on what is happening here.

Reply:
This set of lambdas is finding the associated mission, if any, in which this ship belongs to one of the NPC blocks, using `std::find_if`.